server/registry/names/deployment.go

```diff
@@ -0,0 +1,133 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package names
+
+import (
+	"fmt"
+	"regexp"
+)
+
+// Deployment represents a resource name for an API deployment.
+type Deployment struct {
+	ProjectID    string
+	ApiID        string
+	DeploymentID string
+}
+
+// Validate returns an error if the resource name is invalid.
+// For backward compatibility, names should only be validated at creation time.
+func (d Deployment) Validate() error {
+	r := deploymentRegexp()
+	if name := d.String(); !r.MatchString(name) {
+		return fmt.Errorf("invalid deployment name %q: must match %q", name, r)
+	}
+
+	return validateID(d.DeploymentID)
+}
+
+// Project returns the parent project for this resource.
+func (d Deployment) Project() Project {
+	return d.Api().Project()
+}
+
+// Api returns the parent API for this resource.
+func (d Deployment) Api() Api {
+	return Api{
+		ProjectID: d.ProjectID,
+		ApiID:     d.ApiID,
+	}
+}
+
+// Revision returns an API deployment revision with the provided ID and this resource as its parent.
+func (d Deployment) Revision(id string) DeploymentRevision {
+	return DeploymentRevision{
+		ProjectID:    d.ProjectID,
+		ApiID:        d.ApiID,
+		DeploymentID: d.DeploymentID,
+		RevisionID:   id,
+	}
+}
+
+// Artifact returns an artifact with the provided ID and this resource as its parent.
+func (d Deployment) Artifact(id string) Artifact {
+	return Artifact{
+		name: deploymentArtifact{
+			ProjectID:    d.ProjectID,
+			ApiID:        d.ApiID,
+			DeploymentID: d.DeploymentID,
+			ArtifactID:   id,
+		},
+	}
+}
+
+// Normal returns the resource name with normalized identifiers.
+func (d Deployment) Normal() Deployment {
+	return Deployment{
+		ProjectID:    normalize(d.ProjectID),
+		ApiID:        normalize(d.ApiID),
+		DeploymentID: normalize(d.DeploymentID),
+	}
+}
+
+// Parent returns this resource's parent API resource name.
+func (d Deployment) Parent() string {
+	return d.Api().String()
+}
+
+func (d Deployment) String() string {
+	return normalize(fmt.Sprintf("projects/%s/locations/%s/apis/%s/deployments/%s",
+		d.ProjectID, Location, d.ApiID, d.DeploymentID))
+}
+
+// deploymentCollectionRegexp returns a regular expression that matches a collection of deployments.
+func deploymentCollectionRegexp() *regexp.Regexp {
+	return regexp.MustCompile(fmt.Sprintf("^projects/%s/locations/%s/apis/%s/deployments$",
+		identifier, Location, identifier))
+}
+
+// deploymentRegexp returns a regular expression that matches a deployment resource name.
+func deploymentRegexp() *regexp.Regexp {
+	return regexp.MustCompile(fmt.Sprintf("^projects/%s/locations/%s/apis/%s/deployments/%s$",
+		identifier, Location, identifier, identifier))
+}
```

Comment:
Since these are unexported, consider moving them into package level vars similar to deployment revisions. I don't think there's any benefit to reconstructing the pattern every time we use it.

Reply:
Agreed! I'll make that change after the current set of PRs settle.